src/Microsoft.Health.Api/Features/ByteCounter/ByteCountingStream.cs

```diff
@@ -0,0 +1,135 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using EnsureThat;
+
+namespace Microsoft.Health.Api.Features.ByteCounter
+{
+    /// <summary>
+    /// A stream to wrap an underlying stream and counts the number of bytes passed while operating with it.
+    /// </summary>
+    public class ByteCountingStream : Stream
+    {
+        private readonly Stream _stream;
+        private long _writtenByteCount;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ByteCountingStream"/> class.
+        /// </summary>
+        /// <param name="stream">An underlying stream.</param>
+        public ByteCountingStream(Stream stream)
+        {
+            EnsureArg.IsNotNull(stream, nameof(stream));
+
+            _stream = stream;
+        }
+
+        /// <inheritdoc />
+        public override bool CanRead => _stream.CanRead;
+
+        /// <inheritdoc />
+        public override bool CanSeek => _stream.CanSeek;
+
+        /// <inheritdoc />
+        public override bool CanWrite => _stream.CanWrite;
+
+        /// <inheritdoc />
+        public override long Length => _stream.Length;
+
+        /// <inheritdoc />
+        public override long Position
+        {
+            get { return _stream.Position; }
+            set { _stream.Position = value; }
+        }
+
+        /// <summary>
+        /// The number of bytes that have been written.
+        /// </summary>
+        public long WrittenByteCount { get => _writtenByteCount; }
+
+        /// <inheritdoc />
+        public override void Flush()
+        {
+            _stream.Flush();
+        }
+
+        /// <inheritdoc />
+        public override Task FlushAsync(CancellationToken cancellationToken)
+        {
+            return _stream.FlushAsync(cancellationToken);
+        }
+
+        /// <inheritdoc />
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            return _stream.Read(buffer, offset, count);
+        }
+
+        /// <inheritdoc />
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            return _stream.Seek(offset, origin);
+        }
+
+        /// <inheritdoc />
+        public override void SetLength(long value)
+        {
+            _stream.SetLength(value);
+        }
+
+        /// <inheritdoc />
+        public override void WriteByte(byte value)
+        {
+            _stream.WriteByte(value);
+            Interlocked.Increment(ref _writtenByteCount);
```

Comment:
I appreciate the thread-safety, but `Stream` but reading and writing are not guaranteed to be thread-safe. Instead, users should only have at most a single thread reading from and a single thread writing to a `Stream` at a time. Otherwise, they need to synchronize their access. `Stream` even provides a static method for synchronizing called `Stream.Synchronized(Stream)` for helping with this.

So I don't think you need to worry about using `Interlocked` to increment this count.

Reply:
Ah thanks. We had talked about thread safety in the other PR but I didn't know this about streams.

Reply:
Updated